### ISSUE TITLE
Segfault and assert fail when calculating alignment using MyersHirschberg method

### ISSUE THREAD

**Issue by Martinsos** [seqan/seqan#1711]
Hi Seqan team! First of all, I really like your library - great documentation and easy to use.

I am an author of https://github.com/Martinsos/edlib - library that implements Myers+Hirschberg algorithm in order to calculate Levenshtein distance and find alignment of two sequences. I am trying to compare Edlib with Seqan's implementation of Myers+Hirschberg and pairwise alignment in general, however Seqan's finding of alignment in most cases fails with error - I would appreciate any help on this!

I do initialization like this:
```
TSequence querySeqAn;
for (int idx = 0; idx < querySize; idx++) {
    seqan::appendValue(querySeqAn, query[idx]);  // query is array of unsigned char.
}
TSequence targetSeqAn;
for (int idx = 0; idx < targetSize; idx++) {
    seqan::appendValue(targetSeqAn, target[idx]);  // target is array of unsigned char.
}
TAlign align;
seqan::resize(seqan::rows(align), 2);
seqan::assignSource(seqan::row(align, 0), targetSeqAn);
seqan::assignSource(seqan::row(align, 1), querySeqAn);
```

I am working with genome data, so alphabet is ACTG.

When I search just for score, it is all fine:
```
score = seqan::globalAlignmentScore(querySeqAn, targetSeqAn, seqan::MyersBitVector());
```
works fine.
```
score = seqan::globalAlignmentScore(targetSeqAn, querySeqAn,
                                                        seqan::Score<int, seqan::Simple>(0, -1, -1),
                                                        seqan::AlignConfig<true, false, false, true>(),
                                                        seqan::LinearGaps());
```
also works fine.

However, when I search for alignment like this:
a)
```
int score = seqan::globalAlignment(align, seqan::MyersHirschberg());
```
or like this:
b)
```
score = seqan::globalAlignment(align, seqan::Score<int, seqan::Simple>(0, -1, -1),
                                                   seqan::AlignConfig<true, false, false, true>(),
                                                   seqan::LinearGaps());
```
I get errors!

It does work for some basic, short dummy sequences, but it fails for almost all of the large ones (>15kbp). Errors are sometimes segfault, sometimes it is assert, and sometimes it is something else. They usually do not happen immediately, but after seqan has been running for some itme.

Some examples of error:
1. when comparing two sequences where each has size of about 100k and using code under a), I got `munmap_chunk(): invalid pointer.`
2. when comparing two seqs where first is 1kbp and another is 5000kbp using code b), I got `malloc.c:3723: _int_malloc: Assertion (unsigned long) (size) >= (unsigned long) (nb)' failed.`

Am I doing something wrong on my side, or is there some limit on size of the sequences? Thank you for any help!

**Comment by rrahn** [comment]
First, thanks for reporting to us and that you are happy with our library. 👍 

AFAICS, the alignment invocation looks good.
So it seems to be a bug. 
Unfortunately, we switched to using the ```Finder``` (http://docs.seqan.de/seqan/master/?p=MyersPattern) interface for using the Myer's Bitvector algorithm, but we don't compute the alignment information there, so it might not be sufficient for your comparisons?
We'll investigate the issue.

**Comment by Martinsos** [comment]
@rrahn thank you for your response!

Could a problem be in me allocating align structure on stack?

Thanks for pointing out the Finder, I did not know about that - do you have any code example? How would I do the same call as above? Does this mean that old way of using MyersHirshcberg is not supported any more?

Is there any way I can call MyersHirschberg alignment then?

THanks!

**Comment by rrahn** [comment]
I don't think the stack will be a problem here.
You could print the full stack trace. That might help, to narrow it down.

The Finder interface does not implement MyersHirschberg. Only Myer's with the Ukkonen trick.
Here a small example that should work (I did not test it):

```cpp
#include <seqan/find.h>

using namespace seqan;
...

DnaString needle = "AGTACGAGCATCG"
DnsString hystk = "ATACGACTACTGAGCTA"

Pattern<DnaString, MyersUkkonenGlobal> p(needle);
Finder<DnaString> f(hyskt);
if (find(f, p, -5))
    std::cout << "Found pattern with score: " << getScore(p) << std::endl;
else
    std::cout << "Didn't find pattern!" << std::endl; 
```

**Comment by Martinsos** [comment]
Awesome, thanks - it will give me good idea how to start.

I hope you don't mind if I ask few more questions to clarify the situation - since I am doing comparison I want to use Seqan in the best way possible.

My library implements MyersHirschberg with Ukkonen trick to solve pairwise alignment. It works for [global alignment](https://seqan.readthedocs.io/en/seqan-v2.1.0/Tutorial/PairwiseSequenceAlignment.html#global-alignments) and 2 types of [overlap alignment](https://seqan.readthedocs.io/en/seqan-v2.1.0/Tutorial/PairwiseSequenceAlignment.html#overlap-alignments). When I say "solve", it means it can either return just score or it can return both score and alignment path.

I was looking for analog functionalities in Seqan so I can do the comparison, and I found this:

1. Regarding the global alignment, Seqan has `globalAlignmentScore(querySeqAn, targetSeqAn, MyersBitVector())` for just score and has `globalAlignment(align, MyersHirschberg())` for score and alignment. However, the latter seems to be crashing in most cases. Is this going to be deprecated because of Finder interface?
2. Regarding the first type of overlap alignment, which is defined as `AlignConfig<true, false, false, true>` by Seqan, I found that Seqan has no Myers implementation, but I could use `globalAlignmentScore(targetSeqAn, querySeqAn, Score<int, Simple>(0, -1, -1), AlignConfig<true, false, false, true>(), LinearGaps())` for just score or `globalAlignment(align, Score<int, Simple>(0, -1, -1), AlignConfig<true, false, false, true>(), LinearGaps())` for both score and alignment. Latter often crashes.
However, after you informed me about the Finder interface, it seems I can use Myers implementation for it through Finder.
3. Regarding the second type of overlap alignment, which is defined as `AlignConfig<false, false, false, true>` by Seqan, I found that Seqan has no Myers implementation, but I could use `globalAlignmentScore(targetSeqAn, querySeqAn, Score<int, Simple>(0, -1, -1), AlignConfig<false, false, false, true>(), LinearGaps())` for just score or `globalAlignment(align, Score<int, Simple>(0, -1, -1), AlignConfig<false, false, false, true>(), LinearGaps())` for both score and alignment. Latter often crashes.
However, after you informed me about the Finder interface, it seems I can use Myers implementation through Finder, by specifying FindPrefix specialization tag?

**Comment by rrahn** [comment]
> Regarding the global alignment, Seqan has globalAlignmentScore(querySeqAn, targetSeqAn, MyersBitVector()) for just score and has globalAlignment(align, MyersHirschberg()) for score and alignment. However, the latter seems to be crashing in most cases. Is this going to be deprecated because of Finder interface?

You are right, the first interface does not compute an alignment. This interface can be replaced with the mentioned Finder interface. The second version - the alignment computation - is not replaceable with the Finder interface. Only the Infix version can also (http://docs.seqan.de/seqan/develop/?p=MyersUkkonen) compute an alignment, but not in the way you would think. The finder interface is meant for small read verification. If the alignment is ought to be computed I think a full DP is used to find the beginning of the match. (So it only works as Semi-global alignment, with small query sequences.)

> Regarding the first type of overlap alignment (infix), which is defined as AlignConfig<true, false, false, true> by Seqan, I found that Seqan has no Myers implementation, but I could use globalAlignmentScore(targetSeqAn, querySeqAn, Score<int, Simple>(0, -1, -1), AlignConfig<true, false, false, true>(), LinearGaps()) for just score or globalAlignment(align, Score<int, Simple>(0, -1, -1), AlignConfig<true, false, false, true>(), LinearGaps()) for both score and alignment. Latter often crashes. However, after you informed me about the Finder interface, it seems I can use Myers implementation for it through Finder. 

Unfortunately, the current MyersHirschberg variant does not support the AlignConfig option. Only the standard DP like Needleman-Wunsch and Gotoh algorithms. So here you would have an advantage over our implementations. We can see if we might add at some point. 
Here you are referring to a semi-global alignment, which is possible with the finder interface, although that only the score based method would work here.

> Regarding the second type of overlap alignment (prefix), which is defined as AlignConfig<false, false, false, true> by Seqan, I found that Seqan has no Myers implementation, but I could use globalAlignmentScore(targetSeqAn, querySeqAn, Score<int, Simple>(0, -1, -1), AlignConfig<false, false, false, true>(), LinearGaps()) for just score or globalAlignment(align, Score<int, Simple>(0, -1, -1), AlignConfig<false, false, false, true>(), LinearGaps()) for both score and alignment. Latter often crashes. However, after you informed me about the Finder interface, it seems I can use Myers implementation through Finder, by specifying FindPrefix specialization tag?

Ok, ah, now I see! 😯 
I think here is a misunderstanding about the term LinearGaps here. 
This is not yet a public interface, rather something that is used internally. (That's why it is not documented) 
So the tag you can specify to the alignment algorithms is something like ```NeedlemanWunsch``` or ```Gotoh```, or ```Hirschberg```.
Please see here: http://docs.seqan.de/seqan/develop/?p=AlignmentAlgorithmTags which tags are possible.
The thing is, that LinearGaps, does not mean linear memory. 
That would be achieved by the Hirschberg variant: 
But these algorithms are implemented separately from the rest of the DP code, which was modernised some time ago and were not ported to the new DP engine yet.
You can see what kind of algorithms you can use, and what there limitations are on this site: http://seqan.readthedocs.io/en/master/Tutorial/Algorithms/Alignment/PairwiseSequenceAlignment.html#specialized-alignments

So to summarise it:
* globalAlignmentScore does only compute the score in **linear** space (just one DP column necessary).
* globalAlingment computes a normal alignment with quadratic space for the traceback matrix (probably the main cause of you error, because you allocate to much memory for large alignments).
* globalAlignment with Hirschberg tag, can only compute with linear gap costs and does not allow free-end gaps
* globalAlignment with MyersHirschberg, has the same restrictions as above, but obviously only works with Levenshtein distance.
* globalAlignment with Myers tag, only computes the score with Levenshtein distance.
* Find interface, is an alternative for the alignment interface, and allows global and semi-global alignments but only the score version, without an alignment.

**Comment by Martinsos** [comment]
Thanks @rrahn, I got it now I think! I actually was not confused by linear gaps - I do understand that it does not have to do anything with how much space is consumed for alignment - however I did not know that I can use Hirschberg in this scenario!

Ok, so basically I can not expect much from alignment for any of the scenarios I mentioned (because it has errors for MyersHirschberh and it works only in quadratic space for other two scenarios).

And regarding score, I can use align + MyersBitVector for global, and for semi-global I can use Finder interface -> so the part with Finder interface is new for me, because I was only using align so far. This is great, however i did not get much luck running it! I am getting scores that are not the same as with align or my method (my method and seqan's align do give same results so I believe that is ok).
Could you help me with this, what am I doing wrong?

I was expecting these two to be the same, however I get different results:
```
score = seqan::globalAlignmentScore(targetSeqAn, querySeqAn,
        seqan::Score<int, seqan::Simple>(0, -1, -1),
        seqan::AlignConfig<true, false, false, true>(),
        seqan::LinearGaps());
```
and
```
seqan::Pattern<TSequence, seqan::MyersUkkonen> pattern(querySeqAn);
seqan::Finder<TSequence> finder(targetSeqAn);
seqan::find(finder, pattern);
score = seqan::getScore(pattern);
```

Both should be infix search with levenshtain distance, however I am getting different resutls! Am I doing something wrong when using finder?
Thank you.

**Comment by rrahn** [comment]
Can you give me two sequences for which the MyersHirschberg fails?
I run it with two sequences ~100 Kbp, without problems.
But you mentioned it does not crash for all instances?

Note, that the finder interface, works a little bit different than the alignment. 
You need to specify an error threshold. The Finder finds than all semi-global alignments within this error threshold.
Like so:
```cpp
Pattern<DnaString, MyersUkkonen> pattern(seqV);
Finder<DnaString> finder(seqH);

 while (find(finder, pattern, -100))
     std::cout << "Finder-Score: " << getScore(p) << std::endl;
```
So you need to know already how many errors you want to allow.
This is different to the alignment interface which just finds one best semi-global alignment.

**Comment by Martinsos** [comment]
@rrahn Here are two sequences that cause the crash for global alignment using MyersHirschberg: [seq1](https://www.dropbox.com/s/s5zx65x8o9gsgle/e_coli_536_prefix1000.fasta?dl=0) and [seq2](https://www.dropbox.com/s/s5zx65x8o9gsgle/e_coli_536_prefix1000.fasta?dl=0).

To run this alignment, I used `seqan::globalAlignment(align, seqan::MyersHirschberg())`. I get segmentation fault. It also happens for many other sequences! I managed to get it working only for two really small, dummy sequences. Maybe I am using something wrong?

This is how I create sequences and align object:
```
    TSequence querySeqAn;
    for (int idx = 0; idx < (*querySequences)[0].size(); idx++) {
        seqan::appendValue(querySeqAn, (*querySequences)[0][idx]);
    }
    TSequence targetSeqAn;
    for (int idx = 0; idx < (*targetSequences)[0].size(); idx++) {
        seqan::appendValue(targetSeqAn, (*targetSequences)[0][idx]);
    }
    TAlign align;
    seqan::resize(seqan::rows(align), 2);
    seqan::assignSource(seqan::row(align, 0), targetSeqAn);
    seqan::assignSource(seqan::row(align, 1), querySeqAn);
```

Thank you for explaining the finder interface! Ok, so I need to specify an error threshold. I actually did do that, and I set threshold to big enough value, however I am still getting weird results.
For example, for those two sequences above, correct infix semi-global alignment gives score of 485, while finder interface gives score 984 (I set threshold to 1000)!

Here is how I am using it with those two sequences constructed as I showed above:
```
                    seqan::Pattern<TSequence, seqan::MyersUkkonen> pattern(querySeqAn);
                    seqan::Finder<TSequence> finder(targetSeqAn);
                    seqan::find(finder, pattern, -1000);
                    if (seqan::find(finder, pattern))
                        cout << "\nFound pattern with score: " << seqan::getScore(pattern) << endl;
                    else
                        cout << "\nDidn't find pattern!" << endl;
```

And here is how I defined type `TSequence`: `typedef seqan::String<unsigned char> TSequence;` -> I took this from your tutorial.

**Comment by Martinsos** [comment]
I tried allocating all of seqan data on the heap (sequences and align object), but it did not help - segmentation fault is still happening for finding global alignment path.

I also tried using DnaString instead of the TSequence typedef that I was using, but then I got `malloc.c:3723: _int_malloc: Assertion `(unsigned long) (size) >= (unsigned long) (nb)' failed.` when trying to find global alignment path. And for finder interface, I got the same result as before, 984, which is not correct.

**Comment by rrahn** [comment]
Mhmm, ok why are you not straight assigning the ```(*querySequence)[0]``` to ```querySeqan``` string?
Although, I think it should work the way you constructed it. 
I will use your two sequences and see what happens than.

```cpp
    typedef typename std::remove_reference<decltype(*querySequences[0])>::type TSequence;
    typedef seqan::Align<TSequence, seqan::ArrayGaps> TAlign;

    TAlign align;
    seqan::resize(seqan::rows(align), 2);
    seqan::assignSource(seqan::row(align, 0), (*querySequences)[0]);
    seqan::assignSource(seqan::row(align, 1), (*querySequences)[1]);
```

Are you sure about the score? If it is exactly the same example you run in your experiment, than it seems to me you are aligning one sequence against itself. 
So I would expect the score of the finder to be correct?

I will have a look, but I am not in the office next week. 
So I hope I can look into it today.

**Comment by Martinsos** [comment]
@rrahn I did not know that I can straight assign it! I can try that later, but as you said, it should not be a problem (I am getting correct score when not using finder interface, so I do not think that is a problem).

Hm, I double checked but I am pretty sure I am comparing query with target, and not with itself. The code is exactly as I provided in comment above! Why do you think I am aligning one sequence against itself?

Thanks for the effort!

**Comment by rrahn** [comment]
Ah, sorry, I overlooked that you have two different sequence containers for target and queries.
My bad. So it should be ok than.
Sorry.

**Comment by Martinsos** [comment]
@rrahn no problem! Looking forward to your answer :).

**Comment by rrahn** [comment]
@Martinsos can you check if #1758 fixes the issues?

**Comment by rrahn** [comment]
Fixed by #1758.
@Martinsos Please reopen if you experience still some issues.

**Comment by Martinsos** [comment]
@rrahn thanks, this fixes error with global alignment not working for MyersHirschberg! I would like to warn you - you have left a debugging output in the release version 2.2.0, when doing MyersHirschberg alignment, so a lot of binary values gets printed - you will probably want to remove this.

**Comment by rrahn** [comment]
@Martinsos you are right, we fixed it for the current release.
Thank you.
